From 3ac5fb92954f95c3327a3e201b565eef808468f7 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 05 Aug 2018 22:23:10 +0300
Subject: [PATCH] board: habana: add support for fw flashing

Add supoprt for flashing and checking the firmware with help of
the fw_loader utility.

Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/board/habanalabs/common/local_install_packages b/board/habanalabs/common/local_install_packages
index e6f8c0f..ef84496 100755
--- a/board/habanalabs/common/local_install_packages
+++ b/board/habanalabs/common/local_install_packages
@@ -2,6 +2,7 @@
 
 pre_pkg_lst="protobuf-compiler libprotoc-dev dkms"
 habana_pkg_lst="habanalabs habanatools libhabana"
+fw_ldr=$HOME/fw_loader
 
 function local_msg() {
 	local msg=$1
@@ -24,6 +25,24 @@
 	sudo apt install ${habana_pkg_lst} -y	|| return 1;
 }
 
+function local_program_flash() {
+	local hl_vid=1da3
+	local hl_did=0001
+	local hl_slot=$(lspci -Dd ${hl_vid}:${hl_did} | cut -d\  -f1)
+	local pci_rescan=/sys/bus/pci/rescan
+	local hl_remove=/sys/bus/pci/devices/${hl_slot}/remove
+	local hl_syspath=/sys/class/habanalabs
+
+	local_msg "***************** Programming SPI flash ****************"
+	sudo $fw_ldr || return 1;
+	sleep 5;
+	sudo bash -c "echo 1 > $hl_remove" || return 1;
+	sleep 5;
+	sudo bash -c "echo 1 > $pci_rescan" || return 1;
+	local_msg "************** New SPI flash FW versions ***************"
+	sleep 5;
+	cat $hl_syspath/hl*/*_ver
+}
 
 local_remove_packages
 local_install_packages
@@ -33,3 +52,9 @@
 	exit 3;
 fi
 
+local_program_flash
+if [ $? -ne 0 ]; then
+        local_msg "************** Failed programming flash ****************"
+        local_remove_packages
+        exit 3;
+fi
diff --git a/board/habanalabs/common/remote_load_firmware b/board/habanalabs/common/remote_load_firmware
index 93f5611..e2bb15c 100755
--- a/board/habanalabs/common/remote_load_firmware
+++ b/board/habanalabs/common/remote_load_firmware
@@ -6,6 +6,7 @@
 fw_dir=/lib/firmware/habanalabs/dali
 local_scr_dir=/var/lib/tftpboot/tests
 scr_name=local_install_packages
+fw_ldr=fw_loader
 remote_scr_dir="."
 
 
@@ -22,6 +23,7 @@
 	# sync the latest built binaries
 	for ((j=0; j<3; j++)); do
 		sshpass -p123456 scp ${fw_dir}/* root@${host}:${fw_dir}/	|| break;
+		sshpass -p123456 scp ${local_scr_dir}/${fw_ldr} ${host}:${remote_scr_dir}/	|| break;
 		sshpass -p123456 scp ${local_scr_dir}/${scr_name} ${host}:${remote_scr_dir}/	|| break;
 		return 0;
 	done
